contracts/FlashRollover.sol

```diff
@@ -0,0 +1,344 @@
+    ILoanCore public immutable legacyLoanCore;
+    IOriginationController public immutable originationController;
+    IRepaymentController public immutable legacyRepaymentController;
+    IRepaymentController public immutable repaymentController;
+    IERC721 public immutable borrowerNote;
+    IERC721 public immutable lenderNote;
+    IERC721 public immutable legacyBorrowerNote;
+    IERC721 public immutable legacyLenderNote;
+    IERC721 public immutable assetWrapper;
+    IFeeController public immutable feeController;
+
+    constructor(
+        ILendingPoolAddressesProvider _addressesProvider,
+        ILoanCore _loanCore,
+        ILoanCore _legacyLoanCore,
+        IOriginationController _originationController,
+        IRepaymentController _repaymentController,
+        IRepaymentController _legacyRepaymentController,
+        IERC721 _borrowerNote,
+        IERC721 _legacyBorrowerNote,
+        IERC721 _lenderNote,
+        IERC721 _legacyLenderNote,
+        IERC721 _assetWrapper,
+        IFeeController _feeController
+    ) {
+        ADDRESSES_PROVIDER = _addressesProvider;
+        LENDING_POOL = ILendingPool(_addressesProvider.getLendingPool());
+        loanCore = _loanCore;
+        legacyLoanCore = _legacyLoanCore;
+        originationController = _originationController;
+        repaymentController = _repaymentController;
+        legacyRepaymentController = _legacyRepaymentController;
+        borrowerNote = _borrowerNote;
+        legacyBorrowerNote = _legacyBorrowerNote;
+        lenderNote = _lenderNote;
+        legacyLenderNote = _legacyLenderNote;
+        assetWrapper = _assetWrapper;
+        feeController = _feeController;
+    }
+
+    function rolloverLoan(
+        bool isLegacy,
+        uint256 loanId,
+        LoanLibrary.LoanTerms calldata newLoanTerms,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external override {
+        // Get loan details
+        LoanLibrary.LoanData memory loanData;
+        if (isLegacy) {
+            loanData = legacyLoanCore.getLoan(loanId);
+            uint256 borrowerNoteId = loanData.borrowerNoteId;
+
+            address borrower = legacyBorrowerNote.ownerOf(borrowerNoteId);
+            require(borrower == msg.sender, "Rollover: borrower only");
+        } else {
+            loanData = loanCore.getLoan(loanId);
+            uint256 borrowerNoteId = loanData.borrowerNoteId;
+
+            address borrower = borrowerNote.ownerOf(borrowerNoteId);
+            require(borrower == msg.sender, "Rollover: borrower only");
+        }
+
+        LoanLibrary.LoanTerms memory terms = loanData.terms;
+        uint256 amountDue = terms.principal.add(terms.interest);
+
+        require(newLoanTerms.payableCurrency == terms.payableCurrency, "Currency mismatch");
+        require(newLoanTerms.collateralTokenId == terms.collateralTokenId, "Collateral mismatch");
+
+        uint256 startBalance = IERC20(terms.payableCurrency).balanceOf(address(this));
```

Comment:
So does this mean that the contract can carry a balance? How and why? Do we need some recovery function for stuck funds?

Alternatively you could just make it an invariant that it shouldn't, and you can verify at the end that the balance is 0